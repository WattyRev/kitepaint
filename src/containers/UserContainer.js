import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { getUser, getUserRecognition } from "../redux/modules/user";
import { SET_RECOGNIZED_USER, LOG_IN, LOG_OUT } from "../redux/actions";

/**
 * Maintains a local storage variable that indicates if the user is recognized as a registered user.
 * TODO this needs to be done in redux, because there will need to be more than one of these, and
 * they need to stay in sync.
 *
 * @param {Function} children
 */
class UserContainer extends React.Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    onSetRecognition: PropTypes.func.isRequired,
    onLogIn: PropTypes.func.isRequired
  };

  /**
   * Toggles the variable in state and in local storage
   */
  toggleRecognition = () => {
    const { isRecognizedUser } = this.props;
    this.props.onSetRecognition(!isRecognizedUser);
  };

  render() {
    const {
      firstName,
      id,
      isLoggedIn,
      isLoggingIn,
      lastName,
      username
    } = this.props.user;
    return this.props.children({
      actions: {
        toggleRecognition: this.toggleRecognition,
        logIn: this.props.onLogIn,
        logOut: this.props.onLogOut
      },
      props: {
        firstName,
        id,
        isLoggedIn,
        isLoggingIn,
        isRecognizedUser: this.props.isRecognizedUser,
        lastName,
        username
      }
    });
  }
}

const mapStateToProps = state => ({
  user: getUser(state),
  isRecognizedUser: getUserRecognition(state)
});

const mapDispatchToProps = {
  onLogIn: LOG_IN,
  onLogOut: LOG_OUT,
  onSetRecognition: SET_RECOGNIZED_USER
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UserContainer);
