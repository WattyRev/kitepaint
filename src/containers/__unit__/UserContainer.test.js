import React from "react";
import { shallow, mount } from "enzyme";
import { UserContainer } from "../UserContainer";

describe("UserContainer", () => {
  let props;
  beforeEach(() => {
    props = {
      isRecognizedUser: false,
      user: {
        firstName: "frank",
        id: "abc",
        isLoggedIn: true,
        isLoggingIn: false,
        lastName: "sinatra",
        username: "frankyboi"
      },
      onLogOut: jest.fn(),
      onSetRecognition: jest.fn()
    };
  });
  it("renders", () => {
    expect.assertions(1);
    const wrapper = shallow(
      <UserContainer {...props}>
        {() => {
          <div>Test</div>;
        }}
      </UserContainer>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it("provides the expected props", () => {
    expect.assertions(1);
    const children = jest.fn(() => <div>Test</div>);
    mount(<UserContainer {...props}>{children}</UserContainer>);
    const providedProps = children.mock.calls[0][0].props;
    expect(providedProps).toEqual({
      firstName: "frank",
      id: "abc",
      isLoggedIn: true,
      isLoggingIn: false,
      lastName: "sinatra",
      username: "frankyboi",
      isRecognizedUser: false
    });
  });
  it("provides the expected actions", () => {
    expect.assertions(1);
    const children = jest.fn(() => <div>Test</div>);
    mount(<UserContainer {...props}>{children}</UserContainer>);
    const providedActions = children.mock.calls[0][0].actions;
    const actionKeys = Object.keys(providedActions);
    expect(actionKeys).toEqual(["logOut", "toggleRecognition"]);
  });
  describe("#toggleRecognition", () => {
    it("should call onSetRecognition with false if it isRecognizedUser was true", () => {
      expect.assertions(1);
      props.isRecognizedUser = true;
      const subject = new UserContainer(props);
      subject.toggleRecognition();
      expect(props.onSetRecognition.mock.calls[0][0]).toEqual(false);
    });
    it("should call onSetRecognition with true if it isRecognizedUser was false", () => {
      expect.assertions(1);
      props.isRecognizedUser = false;
      const subject = new UserContainer(props);
      subject.toggleRecognition();
      expect(props.onSetRecognition.mock.calls[0][0]).toEqual(true);
    });
  });
});
