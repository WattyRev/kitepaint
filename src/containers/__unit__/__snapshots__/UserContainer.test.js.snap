// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserContainer renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserContainer
    isRecognizedUser={false}
    onLogOut={[MockFunction]}
    onSetRecognition={[MockFunction]}
    user={
      Object {
        "firstName": "frank",
        "id": "abc",
        "isLoggedIn": true,
        "isLoggingIn": false,
        "lastName": "sinatra",
        "username": "frankyboi",
      }
    }
  >
    [Function]
  </UserContainer>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [
    undefined,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
